exporters/memory/include/opentelemetry/exporters/memory/in_memory_metric_exporter.h

```diff
@@ -0,0 +1,79 @@
+#include "opentelemetry/sdk/metrics/instruments.h"
+#include "opentelemetry/sdk/metrics/push_metric_exporter.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace exporter
+{
+namespace memory
+{
+
+using InMemoryMetricData = InMemoryData<sdk::metrics::ResourceMetrics>;
+
+/// A Push Metric Exporter which accumulates metrics data in memory and allows it to be inspected.
+/// It is not thread-safe.
+class InMemoryMetricExporter final : public sdk::metrics::PushMetricExporter
+{
+  using AggregationTemporalityMap =
+      std::map<sdk::metrics::InstrumentType, sdk::metrics::AggregationTemporality>;
+
+public:
+  /// @param buffer_size a required value that sets the size of the CircularBuffer
+  /// @param temporality Output temporality as a function of instrument kind.
+  InMemoryMetricExporter(size_t buffer_size,
+                         const sdk::metrics::AggregationTemporalitySelector &temporality)
+      : data_(std::make_shared<InMemoryMetricData>(buffer_size)), temporality_(temporality)
+  {}
+
+  InMemoryMetricExporter(const InMemoryMetricExporter &)  = delete;
+  InMemoryMetricExporter(const InMemoryMetricExporter &&) = delete;
+  void operator=(const InMemoryMetricExporter &)          = delete;
+  void operator=(const InMemoryMetricExporter &&)         = delete;
+  ~InMemoryMetricExporter() override                      = default;
+
+  sdk::common::ExportResult Export(const sdk::metrics::ResourceMetrics &data) noexcept override;
+  sdk::metrics::AggregationTemporality GetAggregationTemporality(
+      sdk::metrics::InstrumentType instrument_type) const noexcept override;
+  bool ForceFlush(
+      std::chrono::microseconds timeout = (std::chrono::microseconds::max)()) noexcept override;
+  bool Shutdown(std::chrono::microseconds timeout = std::chrono::microseconds(0)) noexcept override;
+
+  const std::shared_ptr<InMemoryMetricData> &GetData() const;
+
+private:
+  std::shared_ptr<InMemoryMetricData> data_;
+  std::atomic<bool> is_shutdown_{false};
+  sdk::metrics::AggregationTemporalitySelector temporality_;
+};
+
+/// A factory for InMemoryMetricExporter
+class InMemoryMetricExporterFactory
+{
+public:
+  /// Create a InMemorySpanExporter with a default buffer size and aggregation
+  /// temporality selector.
+  /// @param [out] data the InMemorySpanData the exporter will write to,
+  ///                   for the caller to inspect
+  /// @param [in] buffer_size number of entries to save in the circular buffer
+  /// @param [in] temporality output temporality as a function of instrument kind
+  static std::unique_ptr<opentelemetry::sdk::metrics::PushMetricExporter> Create(
+      std::shared_ptr<InMemoryMetricData> &data,
+      size_t buffer_size                                              = 100,
+      const sdk::metrics::AggregationTemporalitySelector &temporality = [](auto) {
+        return sdk::metrics::AggregationTemporality::kCumulative;
+      });
```

Comment:
Instead of declaring a single Create method with optional parameters, instead:
* declare a method with 1 parameter
* declare a method with 2 parameters
* declare a method with 3 parameters

Also, more all the implementation to a .cc file, it should not be in the header file.

What does this change ?

By the time the Create method expose an alternate flavor with 4 parameters, the SDK ABI will be unchanged, with only an addition with one more static method.

We are not there yet, but this is required in the long term if we want opentelemetry-cpp to be delivered cleanly in packages, with a stable interface for the SDK.

Reply:
Done.